---
title: "group5_FIHTD_HW2"
author: "Caden McQuillen, Yuchen Sun, Xiang (Jennie) Li"
date: '2023-02-12'
output:
  html_document:
    toc: yes
---


## Libraries
```{r message=FALSE}
# BiocManager::install("safe")
library(tidyverse)
library(safe)
library(ggplot2)
```

## 1. Enrichment Analysis

### Load expression data
```{r}
data <- load("geod33272.rda")
```

### KEGG pathways
```{r}
num_kegg_pathways <- length(kegg)
print(paste0("The total number of pathways: ", num_kegg_pathways))

TCA_cycle_genes <- kegg$`hsa00020 Citrate cycle (TCA cycle)`
num_TCA_cycle_genes <- length(TCA_cycle_genes)

print(paste0("The total number of genes in TCA cycle pathway: " , num_TCA_cycle_genes))
```

### Gene X Pathway matrix
```{r}
# create Gene by Pathway matrix (Genes are columns, Pathway are rows)
Genes_Pathway_Matrix <- matrix(nrow = length(kegg), ncol = length(geod33272$ID$entrezID))
colnames(Genes_Pathway_Matrix) <- geod33272$ID$entrezID
rownames(Genes_Pathway_Matrix) <- names(kegg)

for (i in 1:length(kegg)){
  currentPathway <- unlist(kegg[i])
  genes <- geod33272$ID$entrezID  %in%  currentPathway
  Genes_Pathway_Matrix[i,]  <- genes
}

```

### SAFE (Significance Analysis of Function and Expression) 

Running SAFE: 
```{r, eval = FALSE, message = FALSE}
# log2 gene expression matrix
gene_expression_matrix <- log2(geod33272$x)
rownames(gene_expression_matrix) <- geod33272$ID$entrezID
# activation status vector
activation_status <- geod33272$status

# Number of tests
num_tests<- length(gene_expression_matrix[,1])
# Target p value
p <- 0.05/num_tests
# Number of permutations needed for target p value
num_permutations <- 10*(1/p)

diff_exp <- safe(X.mat = gene_expression_matrix, y.vec = activation_status, C.mat = t(Genes_Pathway_Matrix),  Pi.mat = num_permutations, error = "FWER.Bonf")

# Takes forever to run safe, so just load the results instead of rerunning 
save(gene_expression_matrix, activation_status, num_tests, p, num_permutations, diff_exp, file = "Safe.Rdata")
```


Top 10 results: 
```{r}
safe_dat <- load("Safe.Rdata")
# top 10
top10 <- safe.toptable(diff_exp,number = 10, pretty = TRUE, description = FALSE)
print(top10)
```


Visualize pathway "hsa03440 Homologous recombination":

```{r}
# plot
safeplot(safe=diff_exp,  cat.name = "hsa03440 Homologous recombination")
```



## 2. Aggeratation-Based Analysis

### Load metabolomics data
The dataset has already been preprocessed by quotient normalization and logged. Metabolites are in columns and samples are in rows of the `dat` data frame. Annotations of the metabolites can be found in annotations and information on gender is stored in variable gender, with 1 = male and 2 = female.
```{r}
data2 <- load("Simulated_metabolomics_data_preprocessed.Rdata")
```

### Eigenmetabolites from PCA and explained variance for Sub_pathway
```{r}
zdat <- scale(dat) # scale dat to have mean=0 and sd=1 for each column (metabolites)
unique_sub_pathways <- unique(annotations$Sub_pathway) # get all subpathways
df_sub <- data.frame(matrix(nrow=length(gender),ncol=length(unique_sub_pathways))) # create empty data frame for pathway representatives
colnames(df_sub) <- unique_sub_pathways # assign colnames as pathway names
expvars_sub <- c()
for (p in unique_sub_pathways) {
  metab <- annotations[annotations$Sub_pathway==p,]$name # identify metabs that belong to pathway
  use_zdat <- zdat[,metab] # get data of the metabs
  pca <- prcomp(use_zdat) # perform pca
  expvar <- (pca$sdev^2)/sum(pca$sdev^2)
  expvars_sub <- c(expvars_sub, expvar[1])
  df_sub[p] <- pca$x[,1] # add first pca coordinate as pathway representative to df
}
dim_sub <- dim(df_sub) # num samples, num variables
print(paste("The number of samples is",dim_sub[1],"and the number of variables is",dim_sub[2]))

expvars_df_sub <- data.frame(expvars_sub)
ggplot(expvars_df_sub,aes(expvars_sub)) + geom_histogram(bins=10)
print(paste("The average explained variance is",mean(expvars_sub)))
```

Some pathways have explained variance of 1 because there's only one metabolite in that pathway.

### Eigenmetabolites from PCA and explained variance for Super_pathway
```{r}
unique_super_pathways <- unique(annotations$Super_pathway) # get all superpathways
df_super <- data.frame(matrix(nrow=length(gender),ncol=length(unique_super_pathways))) # create empty data frame for pathway representatives
colnames(df_super) <- unique_super_pathways # assign colnames as pathway names
expvars_super <- c()
for (p in unique_super_pathways) {
  metab <- annotations[annotations$Super_pathway==p,]$name # identify metabs that belong to pathway
  use_zdat <- zdat[,metab] # get data of the metabs
  pca <- prcomp(use_zdat) # perform pca
  expvar <- (pca$sdev^2)/sum(pca$sdev^2)
  expvars_super <- c(expvars_super, expvar[1])
  df_super[p] <- pca$x[,1] # add first pca coordinate as pathway representative to df
}
dim_super <- dim(df_super) # num samples, num variables
print(paste("The number of samples is",dim_super[1],"and the number of variables is",dim_super[2]))

expvars_df_super <- data.frame(expvars_super)
ggplot(expvars_df_super,aes(expvars_super)) + geom_histogram(bins=5)
print(paste("The average explained variance is",mean(expvars_super)))
```

The super-pathways all have number of metabolites larger than 1 so there's no explained variance being 1. The average explained variance is a lot higher for sub-pathway analysis, because super-pathway is too broad and you have metabolites potentially being regulated in opposite directions which could mute the overall effect and make it difficult to analyze the difference.


### Pathway differences between males and females

```{r message=FALSE}
library(limma)
```

```{r}
design <- model.matrix(~gender, data=df_sub)
```


```{r}
fit <- lmFit(t(df_sub), design)
```


```{r}
fit <- eBayes(fit)
```

If the coefficient estimate is positive, it means that the mean the value is higher in females compared to males, and vice versa if the estimate is negative.


```{r}
results <- topTable(fit, coef=2, adjust.method="fdr", sort.by="P", number=Inf)

hist(results$logFC, breaks = 50)

de_pathways <- rownames( subset(results, abs(logFC) > 1 & adj.P.Val < 0.05) )
```
```{r}
design2 <- model.matrix(~gender, data=df_sub)
```


```{r}
fit2 <- lmFit(t(df_super), design2)
```


```{r}
fit2 <- eBayes(fit2)
```

If the coefficient estimate is positive, it means that the mean the value is higher in females compared to males, and vice versa if the estimate is negative.


```{r}
results2 <- topTable(fit2, coef=2, adjust.method="fdr", sort.by="P", number=Inf)

hist(results2$logFC, breaks = 50)

de_pathways2 <- rownames( subset(results2, abs(logFC) > 1 & adj.P.Val < 0.05) )
```



```{r}
ggplot(results, aes(x=logFC, y=-log10(adj.P.Val))) +
  geom_point(aes(color=ifelse(abs(logFC) > 1 & adj.P.Val < 0.05, "red", "black")), size=1) +
  theme_bw() +
  scale_color_identity(guide="none") +
  xlab("Log2 Fold Change") +
  ylab("-Log10 Adjusted P-value") +
  ggtitle("Volcano Plot of Differential Expression (Sub Pathways)") +
  theme(plot.title = element_text(hjust = 0.5))
```
```{r}
ggplot(results2, aes(x=logFC, y=-log10(adj.P.Val))) +
  geom_point(aes(color=ifelse(abs(logFC) > 1 & adj.P.Val < 0.05, "red", "black")), size=1) +
  theme_bw() +
  scale_color_identity(guide="none") +
  xlab("Log2 Fold Change") +
  ylab("-Log10 Adjusted P-value") +
  ggtitle("Volcano Plot of Differential Expression (Super Pathways)") +
  theme(plot.title = element_text(hjust = 0.5))
```



```{r fig.height=13, fig.width=8}

results$subpathway <- rownames(results) #add column for subpathways
results$superpathway <- annotations$Super_pathway[match(results$subpathway, annotations$Sub_pathway)]
results$plotcolor <- ifelse(results$logFC > 0, "Female", "Male") #color for enrichment
ggplot(results, aes(x=-log10(adj.P.Val), y=reorder(subpathway, logFC, decreasing = T), label=subpathway)) + 
  geom_point(stat='identity', aes(col =plotcolor), size=3 )  + facet_grid(rows = vars(superpathway), scales="free_y")+
  scale_color_manual(name="Log FC", 
                     labels = c("Female", "Male"), 
                     values = c("Female" = "red", "Male"="blue")) + 
  labs(title="Sub Pathways", y = "Subpathways") + geom_vline(xintercept = 1.30102999566) 
#add line for significance 

```

```{r}

results2$superpathway <- rownames(results2) #add column for subpathways
results2$plotcolor <- ifelse(results2$logFC > 0, "Female", "Male") #color for enrichment
ggplot(results2, aes(x=-log10(adj.P.Val), y = reorder(superpathway, logFC, decreasing = T), label=superpathway)) + 
  geom_point(stat='identity', aes(col =plotcolor), size=3 )  +
  scale_color_manual(name="Log FC", 
                     labels = c("Female", "Male"), 
                     values = c("Female" = "red", "Male"="blue")) + 
  labs(title="Super Pathways", y = "Subpathways") + geom_vline(xintercept = 1.30102999566) #add line for significance 

```

